File: src/oasees_sdk/stack/nce.py

import subprocess
import json
import time
import hashlib

class NCE:
    """
    Node Component Executor (NCE) - Mock Implementation
    
    Responsible for:
    1. Accepting Execution Tickets (Interface M7).
    2. Translating Tickets to K8s Job Manifests.
    3. Applying Manifests (Interface M8).
    4. Monitoring and Cleaning up resources.
    """
    
    def __init__(self, nds):
        self.nds = nds
        print("[NCE] Initialized.")

    def submit_ticket(self, ticket):
        """
        M7: Submit Execution Ticket
        Accepts a fully resolved 'Context Bundle' from NWO.
        """
        job_id = ticket.get("job_id", f"job-{int(time.time())}")
        print(f"[NCE] Received Execution Ticket: {job_id}")
        if ticket.get("target_node"):
            print(f"[NCE] Target node: {ticket.get('target_node')}")
        
        # 1. Translate to Manifest
        manifest = self.generate_manifest(ticket)
        
        # 2. Apply Manifest (M8)
        self.apply_manifest(manifest)
        
        # 3. Update Status
        self.nds.update_status(job_id, "RUNNING")
        
        return job_id

    def generate_manifest(self, ticket):
        """
        Generate a K8s Job manifest from the execution ticket.
        """
        print("[NCE] Generating Kubernetes Job Manifest...")
        # In a real implementation, this would use the templates from `training_workload.py`
        env_vars = ticket.get("env_vars") or {}
        if isinstance(env_vars, dict):
            env_list = [{"name": k, "value": str(v)} for k, v in env_vars.items()]
        else:
            env_list = env_vars

        job_id = ticket.get("job_id")
        service_id = ticket.get("service_id")
        execution_strategy = ticket.get("execution_strategy")
        target_node = ticket.get("target_node")

        if job_id:
            env_list.append({"name": "OASEES_JOB_ID", "value": job_id})
        if service_id:
            env_list.append({"name": "OASEES_SERVICE_ID", "value": service_id})
        if execution_strategy:
            env_list.append({"name": "OASEES_EXEC_STRATEGY", "value": execution_strategy})
        if target_node:
            env_list.append({"name": "OASEES_TARGET_NODE", "value": target_node})

        manifest = {
            "apiVersion": "batch/v1",
            "kind": "Job",
            "metadata": {"name": ticket["job_id"]},
            "spec": {
                "template": {
                    "spec": {
                        "containers": [{
                            "name": "workload",
                            "image": ticket.get("image", "oasees/ml-base-image:latest"),
                            "env": env_list
                        }],
                        "restartPolicy": "Never"
                    }
                }
            }
        }
        return manifest

    def apply_manifest(self, manifest):
        """
        M8: Apply Job Manifest (kubectl apply)
        """
        print(f"[NCE] Applying Manifest: {manifest['metadata']['name']}")
        
        # Verify valid manifest
        if not manifest:
            print("[NCE] Error: Empty manifest generated.")
            return False

        # In a real scenario, we would use `subprocess.run(['kubectl', 'apply', ...])`
        # For this prototype, we print the command.
        print(f"[NCE][MOCK] execute: kubectl apply -f {manifest['metadata']['name']}.yaml")
        time.sleep(1) # Simulate delay
        return True

    def generate_attestation(self, manifest, ticket):
        """
        Create a simple execution attestation based on manifest hash.
        """
        manifest_str = json.dumps(manifest, sort_keys=True)
        manifest_hash = hashlib.sha256(manifest_str.encode()).hexdigest()
        attestation = {
            "job_id": ticket.get("job_id"),
            "node_id": (ticket.get("node_profile") or {}).get("node_id"),
            "manifest_hash": manifest_hash,
            "timestamp": time.time()
        }
        print(f"[NCE] Attestation generated: {attestation['manifest_hash'][:12]}...")
        return attestation

    def cleanup(self, job_id):
        """
        Cleanup resources after completion.
        """
        print(f"[NCE] Cleaning up resources for job: {job_id}")
        self.nds.update_status(job_id, "CLEANED")
        self.nds.release_resources(job_id)

    def monitor_job(self, job_id):
        """
        Simulate basic job monitoring and status transitions.
        """
        print(f"[NCE] Monitoring job: {job_id}")
        time.sleep(0.4)
        self.nds.update_status(job_id, "RUNNING")
        time.sleep(0.6)
        self.nds.update_status(job_id, "SUCCEEDED")
        return {"job_id": job_id, "status": "SUCCEEDED"}

File: src/oasees_sdk/stack/demo-server/lib/nce.js
// NCE - Node Component Executor
// Accepts execution tickets, generates K8s manifests, applies them

const EventEmitter = require('events');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const os = require('os');
const crypto = require('crypto');

class NCE extends EventEmitter {
  constructor(nds) {
    super();
    this.nds = nds;
    this.emit('action', { component: 'NCE', action: 'init', log: '[NCE] Initialized. Component executor ready.' });
  }

  submitTicket(ticket) {
    const jobId = ticket.job_id || `job-${Date.now()}`;
    this.emit('action', {
      component: 'NCE', action: 'submitTicket',
      log: `[NCE] Received Execution Ticket: ${jobId}`,
      data: ticket
    });

    // 1. Generate manifest
    const manifest = this.generateManifest(ticket);

    // 2. Apply manifest (M8)
    const applyResult = this.applyManifest(manifest);

    // 3. Update status
    this.nds.updateStatus(jobId, 'RUNNING');

    return { jobId, manifest, applyResult };
  }

  generateManifest(ticket) {
    this.emit('action', {
      component: 'NCE', action: 'generateManifest',
      log: '[NCE] Generating Kubernetes Job Manifest...'
    });

    const envVars = ticket.env_vars || {};
    const envList = Array.isArray(envVars)
      ? envVars
      : Object.entries(envVars).map(([k, v]) => ({ name: k, value: String(v) }));

    const jobId = ticket.job_id;
    const serviceId = ticket.service_id;
    const execStrategy = ticket.execution_strategy;
    const targetNode = ticket.target_node;

    if (jobId) envList.push({ name: 'OASEES_JOB_ID', value: jobId });
    if (serviceId) envList.push({ name: 'OASEES_SERVICE_ID', value: serviceId });
    if (execStrategy) envList.push({ name: 'OASEES_EXEC_STRATEGY', value: execStrategy });
    if (targetNode) envList.push({ name: 'OASEES_TARGET_NODE', value: targetNode });

    const manifest = {
      apiVersion: 'batch/v1',
      kind: 'Job',
      metadata: { name: ticket.job_id },
      spec: {
        template: {
          spec: {
            containers: [{
              name: 'workload',
              image: ticket.image || 'cognets/ml-base-image:latest',
              env: envList
            }],
            restartPolicy: 'Never'
          }
        }
      }
    };

    this.emit('action', {
      component: 'NCE', action: 'manifestGenerated',
      log: `[NCE] Manifest generated for job: ${ticket.job_id}`,
      data: manifest
    });

    return manifest;
  }

  applyManifest(manifest) {
    const jobName = manifest.metadata.name;
    this.emit('action', {
      component: 'NCE', action: 'applyManifest',
      log: `[NCE][M8] Applying manifest: ${jobName}`
    });

    // Write actual YAML file
    const yaml = this._toYaml(manifest);
    const tmpFile = path.join(os.tmpdir(), `${jobName}.yaml`);
    fs.writeFileSync(tmpFile, yaml);

    this.emit('action', {
      component: 'NCE', action: 'manifestWritten',
      log: `[NCE] Wrote manifest to ${tmpFile}`,
      data: { file: tmpFile, yaml }
    });

    // Try kubectl, fall back gracefully
    let cmdOutput = '';
    let command = '';
    try {
      command = `kubectl apply --dry-run=client -f ${tmpFile}`;
      cmdOutput = execSync(command, { encoding: 'utf-8', timeout: 5000 });
    } catch (e) {
      // kubectl not available — fall back to cat
      command = `cat ${tmpFile}`;
      try {
        cmdOutput = execSync(command, { encoding: 'utf-8', timeout: 3000 });
      } catch (e2) {
        cmdOutput = yaml; // ultimate fallback
      }
      cmdOutput = `[kubectl not available — showing manifest]\n${cmdOutput}`;
    }

    this.emit('action', {
      component: 'NCE', action: 'manifestApplied',
      log: `[NCE] Command executed: ${command}`,
      data: { command, output: cmdOutput }
    });

    // Cleanup temp file
    try { fs.unlinkSync(tmpFile); } catch (_) {}

    return { command, output: cmdOutput };
  }

  cleanup(jobId) {
    this.emit('action', {
      component: 'NCE', action: 'cleanup',
      log: `[NCE] Cleaning up resources for job: ${jobId}`,
      data: { jobId }
    });
    this.nds.updateStatus(jobId, 'CLEANED');
    this.nds.releaseResources(jobId);
  }

  generateAttestation(manifest, ticket) {
    const manifestStr = JSON.stringify(manifest);
    const manifestHash = crypto.createHash('sha256').update(manifestStr).digest('hex');
    const attestation = {
      job_id: ticket.job_id,
      node_id: (ticket.node_profile || {}).node_id,
      manifest_hash: manifestHash,
      timestamp: Date.now() / 1000
    };
    this.emit('action', {
      component: 'NCE', action: 'attestation',
      log: `[NCE] Attestation generated: ${manifestHash.slice(0, 12)}...`,
      data: attestation
    });
    return attestation;
  }

  _toYaml(obj, indent = 0) {
    // Simple JSON-to-YAML converter for K8s manifests
    const pad = '  '.repeat(indent);
    let out = '';
    for (const [key, val] of Object.entries(obj)) {
      if (val === null || val === undefined) continue;
      if (Array.isArray(val)) {
        out += `${pad}${key}:\n`;
        for (const item of val) {
          if (typeof item === 'object') {
            const lines = this._toYaml(item, indent + 2).split('\n').filter(Boolean);
            out += `${pad}- ${lines[0].trim()}\n`;
            for (let i = 1; i < lines.length; i++) out += `${pad}  ${lines[i].trim()}\n`;
          } else {
            out += `${pad}- ${item}\n`;
          }
        }
      } else if (typeof val === 'object') {
        out += `${pad}${key}:\n${this._toYaml(val, indent + 1)}`;
      } else {
        out += `${pad}${key}: ${val}\n`;
      }
    }
    return out;
  }
}

module.exports = NCE;
